### PR TITLE
Refactoring v2 0

### PR DESCRIPTION
**Pull request by osorensen**
Still just a draft. Putting it here to get the tests.

### CODE REVIEW

**Comment by osorensen** [comment]
@wleoncio, happy new year to you!

I think this humongous PR is getting ready for review, with currently with 11,074 insertions and 13,429 deletions.

All kinds of feedback appreciated, and we're not in a rush.

**Comment by wleoncio** [comment]
Happy new year, Øystein!

Good job on the package! Just got back last night from a long vacation. I'm still going through a month's worth of e-mails, but I've already seen quite a few BM issues got resolved. I'll get on this ASAP with the attention it deserves. :)

**Comment by osorensen** [comment]
The WARNING causing `R-CMD-check / ubuntu-latest` to fail is due to Rcpp and not BayesMallows. Same error can be found [here](https://cran.r-project.org/web/checks/check_results_Rcpp.html)

image

**Comment by osorensen** [comment]
The last commit involved some relatively major changes to the C++ code. Here's a comment on the approach.

I originally thought it would be sufficient to use the R `parallel` package to set up parallel processing of SMC particles, but this turned out to be wrong once I sat down and thought about it, at least it's suboptimal. Hence I think I've found a better approach, which is included in [this commit](https://github.com/ocbe-uio/BayesMallows/pull/347/commits/92cbbec51679681e07b4cf052f22323e0c14c593).

Wherever possible, I've replaced a standard for-loop with a call to `std::for_each`, as in this place:

https://github.com/ocbe-uio/BayesMallows/blob/e7eec79c624ddb3b7a2a3579d90ac1266724f8bb/src/smc_augmentation_class.cpp#L73-L104

The results are identical in the current implementation, but we can now allow parallel processing of the loop with the following tweak, supported in C++17 and above:

```cpp
#include <execution>

// define vector v etc

std::for_each(
    std::execution::par,
    v.begin(), v.end(), [](type& t){
    // do something to t
     }
);
```

In one overload of `std::for_each`, we can set the execution policy. It defaults to `std::execution::seq`, but we can also set it to `std::execution::par` and `std::execution::par_unseq`, and this can hopefully lead to speedup.

One caveat is that the `<execution>` header is not implemented in Apple clang, and since this is the compiler used by CRAN for building packages on macOS, doing it as simple as I showed in the snippet above won't work on CRAN. However, I consider two alternatives:

- The best option, to use the `<execution>` header when available and otherwise fall back on the default sequential execution, possibly using `has_include` defined [here](https://en.cppreference.com/w/cpp/preprocessor/include).
- Keep a separate branch on GitHub that uses the `<execution>` header, but let then CRAN package run sequentially for now. If I can't get the first option to work, this might also be fine. The thing is that I need the parallel processing option for the paper I'm writing on SMC algorithms for the Mallows model.

**Comment by wleoncio** [comment]
> One caveat is that the `<execution>` header is not implemented in Apple clang, and since this is the compiler used by CRAN for building packages on macOS, doing it as simple as I showed in the snippet above won't work on CRAN.

So what's currently happening on macOS, does the function falls back to serial computation or does it fail altogether?

**Comment by osorensen** [comment]
> > One caveat is that the `<execution>` header is not implemented in Apple clang, and since this is the compiler used by CRAN for building packages on macOS, doing it as simple as I showed in the snippet above won't work on CRAN.
> 
> So what's currently happening on macOS, does the function falls back to serial computation or does it fail altogether?

Currently I'm not using the `<execution>` header, so it uses the following definition which does not have an execution strategy:

```cpp
template< class InputIt, class UnaryFunction >
UnaryFunction for_each( InputIt first, InputIt last, UnaryFunction f );
```

Hence, build succeeds on macOS and on all other platforms.

File: work-docs/augmentation_proposal.h
Comment (wleoncio) [inline work-docs/augmentation_proposal.h:1]: Why not just delete this file?

Comment (osorensen) [reply work-docs/augmentation_proposal.h:1]: Indeed, will do.

Comment (osorensen) [reply work-docs/augmentation_proposal.h:1]: There are actually lots of files in work-docs that I should remove. Will do a cleaning.

Comment (wleoncio) [reply work-docs/augmentation_proposal.h:1]: Nevermind, I misread the diff. Need a break. 🤕

Comment (osorensen) [reply work-docs/augmentation_proposal.h:1]: :-) I'll clean up that directly anyway

Comment (wleoncio) [reply work-docs/augmentation_proposal.h:1]: Great, I'm all in favor of cleaning. :)

**Review by wleoncio — APPROVED** [review]
That was a long read, but it was worth it, I've learned quite a bit about R there. :)

I'm curious about why the unit tests involving hardcoded values (defined with `structure()`) were removed, though.

Nonetheless, as long as the new tests seem sufficient to detect wrong output in future changes (something I had trouble assessing), this looks good to merge. 👍🏽

**Comment by osorensen** [comment]
> I'm curious about why the unit tests involving hardcoded values (defined with structure()) were removed, though.

Yes, that's a good question. The reason is that they became terribly hard to maintain, since very small changes in the code would mess up the order of the random number generation, and hence making me have to rerun and copy-paste the `structure` for all tests. I have a couple of hard-coded tests still, both for SMC and MCMC, but maintaining all the previous ones just became too hard. The test coverage is still close to 100 %, so I hope we'll be able to catch errors still.

**Pull request merged**